### PR TITLE
Add production Alpine and Debian Stretch Docker images

### PR DESCRIPTION
**Pull request by kaiyou**
This PR is currently for reference mostly, as it will conflict with https://github.com/Chocobozzz/PeerTube/pull/213 anyway. It implements a production Alpine-based image with the following concepts in mind:

- the image is built from source (see https://github.com/Chocobozzz/PeerTube/pull/213#discussion_r162807203 for the discussion)
- the image only contains PeerTube and will run a single process, no caching webserver or frontend
- configuration is committed to ``support/docker/config`` to avoid duplicating a common configuration file
- as few layers as possible while maintaining readability

### CODE REVIEW

File: support/docker/config/production.yaml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+database:
+  hostname: 'db'
+  port: 5432
+  suffix: ''
+  username: 'postgres'
+  password: 'postgres'
+
+# From the project root directory
+storage:
+  avatars: '../data/avatars/'
```
Comment (djmaze) [inline support/docker/config/production.yaml]: So this is `/usr/src/data`? You need to adjust the `VOLUME` directive in the Dockerfile accordingly. (Yes, I made a similar mistake in my PR ;))

Comment (kaiyou) [reply support/docker/config/production.yaml]: This is relative to ``/app``, so it is actually ``/data/avatars``.

Comment (djmaze) [reply support/docker/config/production.yaml]: Ah, sorry for the noise.

**Comment by djmaze** [comment]
I like this. Maybe we should go with this PR for now and try to get automated builds working (probably via Travis?). We can add sample Compose / Swarm configs later.

**Comment by Dryusdan** [comment]
You haven't any problem with npm build ? 

There are too many layer but it's cool  :D

**Comment by kaiyou** [comment]
@Dryusdan npm build uses scripty, which fails with the default Alpine shell, hence the ``bash -c`` call. Also, I am open to suggestions as to limiting the number of layers without impacting readability.

**Comment by Dryusdan** [comment]
@kaiyou yeah ! You unlock me o/

Actually I make this image (github mirror repo) : https://github.com/Dryusdan/docker-peertube/blob/master/Dockerfile.alpine

**Comment by Dryusdan** [comment]
So you can inspired if you would :)

**Comment by djmaze** [comment]
I have to say I really don't get the obsession of people wanting to minimize the amount of layers at all costs. With today's docker graph drivers, I believe the overhead is minimal. 

I'd rather optimize for better layer caching, which allows faster updates in production. (See my PR.)

**Comment by Dryusdan** [comment]
If you build your image on the same host, minimal layer is usefull...
In CI build docker in docker, the multiple layer takes longer to build than little layer because cache doesn't exist for this method

**Comment by djmaze** [comment]
@Dryusdan 

> If you build your image on the same host, minimal layer is usefull...

I suppose you mean "minimal number of layers". So why should it be better when building on the same host? A multi-layer approach will probably be much faster since you can skip building many of the base layers if only the application code was changed.

> In CI build docker in docker, the multiple layer takes longer to build than little layer because cache doesn't exist for this method

Right, with Travis there is no Docker caching currently, although there is an improvement with [`--cache-from`](http://atodorov.org/blog/2017/08/07/faster-travis-ci-tests-with-docker-cache/).

But does it really matter if a CI build takes 5 or 7 minutes? This is not something you impatiently wait for, do you?

**Comment by Dryusdan** [comment]
Use cache is can create some problem with version of paquet, because cache "create" outdated packet (ouch, bad english). The base image first.
Then, in production use image not build image not create or build by trust-worthy machine is not very secure

**Comment by Chocobozzz** [comment]
> I like this. Maybe we should go with this PR for now and try to get automated builds working (probably via Travis?). We can add sample Compose / Swarm configs later.

Maybe @kaiyou @Dryusdan @djmaze you could merge your different work in one repository, and make only one PR? For example this alpine image with @djmaze docker compose or whatever.

**Comment by kaiyou** [comment]
I'll try and merge both tonight. I'll update if I cannot deliver in time.

**Comment by kaiyou** [comment]
I have just merged @Dryudsan commits into the directory structure of this PR. I have not tested yet, so it probably doesn't event build.

**Comment by kaiyou** [comment]
I just tested, it seems both images build successfully. Now some testing :)

**Comment by kaiyou** [comment]
Quick update: the Debian image (stretch based) works fine and has been running in production for two days now. The Alpine image still segfaults and I need time or help to fix it.

I suggest we merge but only configure builds for the Debian image at this time.

**Comment by Chocobozzz** [comment]
@kaiyou Please put your files inside a `/support/docker/production` directory and I'll merge. Thanks for your work! :)

File: support/docker/Dockerfile.alpine [hunk 2]
```diff
@@ -0,0 +1,21 @@
+FROM node:8-alpine
+
+# Install dpeendencies
```
Comment (djmaze) [inline support/docker/Dockerfile.alpine]: - [ ] Small typo

File: support/docker/Dockerfile.stretch [hunk 3]
```diff
@@ -0,0 +1,25 @@
+FROM node:8-stretch
+
+# Install dpeendencies
```
Comment (djmaze) [inline support/docker/Dockerfile.stretch]: - [ ] Small typo

File: support/docker/Dockerfile.stretch [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ && apt-get -y install ffmpeg \
+ && rm /var/lib/apt/lists/* -fR
+
+# Install the application
+WORKDIR /app
+
+COPY . ./
+RUN yarn install --pure-lockfile && npm run build
+
+# Configuration the application
```
Comment (djmaze) [inline support/docker/Dockerfile.stretch]: - [ ] `Configuring`?

**Comment by kaiyou** [comment]
@Chocobozzz I did, thank you @djmaze for the review, I fixed the typos, merged the latest develop, added redis settings accordingly.

**Comment by Chocobozzz** [comment]
Thanks @kaiyou for your awesome work!

**Comment by GammaNu** [comment]
Je viens de tester, soit j'ai : 
```
# docker-compose up app
ERROR: No such service: app
```
soit : 
```
 # docker-compose up
Starting peertube_db_1 ... done
Starting peertube_peertube_1 ... done
Attaching to peertube_db_1, peertube_peertube_1
db_1        | 2018-03-11 13:19:57.272 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
db_1        | 2018-03-11 13:19:57.272 UTC [1] LOG:  listening on IPv6 address "::", port 5432
db_1        | 2018-03-11 13:19:57.338 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
db_1        | 2018-03-11 13:19:57.387 UTC [22] LOG:  database system was shut down at 2018-03-11 13:18:36 UTC
db_1        | 2018-03-11 13:19:57.401 UTC [1] LOG:  database system is ready to accept connections
peertube_1  | npm ERR! path /app/package.json
peertube_1  | npm ERR! code EACCES
peertube_1  | npm ERR! errno -13
peertube_1  | npm ERR! syscall open
peertube_1  | npm ERR! Error: EACCES: permission denied, open '/app/package.json'
peertube_1  | npm ERR!  { Error: EACCES: permission denied, open '/app/package.json'
peertube_1  | npm ERR!   stack: 'Error: EACCES: permission denied, open \'/app/package.json\'',
peertube_1  | npm ERR!   errno: -13,
peertube_1  | npm ERR!   code: 'EACCES',
peertube_1  | npm ERR!   syscall: 'open',
peertube_1  | npm ERR!   path: '/app/package.json' }
peertube_1  | npm ERR!
peertube_1  | npm ERR! Please try running this command again as root/Administrator.
peertube_1  |
peertube_1  | npm ERR! A complete log of this run can be found in:
peertube_1  | npm ERR!     /data/.npm/_logs/2018-03-11T13_19_58_817Z-debug.log
peertube_peertube_1 exited with code 243
```

PS : pour que le build de l'image se passe bien, j'ai du modifier le Dockerfile.stretch en Dockerfile et : 
```
COPY src ./
```
src dans le quel j'ai fait un git clone de peertube pour qu'il y ai bien les package.json et autre yarn.lock pour que la constrution de l'image se passe sans problèmes.

**Comment by kaiyou** [comment]
As-tu essayé de build depuis la racine du projet ?

```
docker build -t peertube:stretch -f support/docker/production/Dockerfile.stretch .
```

**Comment by GammaNu** [comment]
nop, je fait ça tout de suite

**Comment by kaiyou** [comment]
In english for the community to be able to read: the docker image seriously lacks documentation. I'll try to remedy that.

**Comment by GammaNu** [comment]
AS: I've just edited my message to translate them in english.
```
docker build -t peertube:stretch -f support/docker/production/Dockerfile.stretch .
```
works, but it dosent change docker-compose behavior.

**Comment by kaiyou** [comment]
I have just updated the compose file in PR #332. Maybe its best we follow up there and make sure the resulting compose example is working properly

**Comment by Chocobozzz** [comment]
Yes please remove it

**Pull request merged**